workload: enhance querybench

yuzefovich opened the pull request:
Release justification: non-production code changes.

`querybench` workload has been enhanced to support queries that consist
of multiple statements that are present on a single line. Previously,
the workload would fail on such queries because it attempts to `Prepare`
the whole line which would error out. Now we're still attempting to
prepare the whole line, but if that fails, we will store plain query and
will be executing it using `DB.Query` rather that `Stmt.Query`.

Fixes: #46547.
Fixes: #46607.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/46583)
<!-- Reviewable:end -->

asubiotto reviewed (commented):
I think the release justification is wrong, this is a non-production code change, right?

Is it possible to just put all statements on one line and have the tool execute that line?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46583#-:-M3LH-MoD_bv421ZPZqs:bej4j1y)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto)



<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
Updated the tool to support multiple statements on a single line separated by semicolon. Initially, I didn't like this idea because we won't be able to support queries that have `;` as a symbol inside of the query itself, but we're unlikely to hit that problem.

Updated the release justification.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46583#-:-M3N4YAn5tgRwe3ao6f0:b-5bxsv1)* status: :shipit: complete! 0 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

asubiotto reviewed (commented):
I think I'm missing something. Can't we just pass in the whole string as is to the driver? I'm not sure why we have to introduce the idea of query groups at all.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46583#-:-M3QQL5zDbcusCBTdwf0:bvjtqs6)* status: :shipit: complete! 0 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

yuzefovich commented:
The problem is that we're `Prepare`ing the whole line, and if the line contains multiple statements, `Prepare` will fail. I believe we're using prepare to reduce the overhead when benchmarking, but it doesn't work for multi-statement queries as is.

Hm, I guess we could try to prepare the whole line, and if that fails, we would keep the whole line together and execute it using `db.Query`

asubiotto reviewed (commented):
I see, up to you.

> Reviewed 3 of 3 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46583#-:-M3RSMwsBO06BdKBftXZ:b-j1zk7a)* status: :shipit: complete! 0 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

yuzefovich committed:
workload: enhance querybench

Release justification: non-production code changes.

`querybench` workload has been enhanced to support queries that consist
of multiple statements that are present on a single line. Previously,
the workload would fail on such queries because it attempts to `Prepare`
the whole line which would error out. Now we're still attempting to
prepare the whole line, but if that fails, we will store plain query and
will be executing it using `DB.Query` rather that `Stmt.Query`.

Release note: None

yuzefovich commented:
Updated, RFAL.

asubiotto reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 2 of 2 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46583#-:-M3RX-O20vmbvfuC2e8K:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

yuzefovich commented:
TFTR!

bors r+

yuzefovich commented:
bors r+
